File: src/oneShot/testScripters/wordUse/wordUse.py
# wordUse - count words in file and print top ten
import sys

def usage():
    """explain usage and exit"""
    sys.exit("""wordUse - count words in file and print top ten 
usage:
    wordUse files""")

class wordCount:
    """Count number of uses of word."""
    def __init__(self, word):
        self.word = word
	self.count = 0
    def __cmp__(self,other):
        return other.count - self.count

   
def wordUse(file):
    """Count words in file and print top ten"""
    f = open(file)
    dict = {}
    while 1:
        line = f.readline()
	if line == '':
	    break
	words = line.split()
	for word in words:
	    if not dict.has_key(word):
		wc = wordCount(word)
		dict[word] = wc
	    else:
	        wc = dict[word]
	    wc.count += 1;
    list = []
    for key in dict.keys():
        list.append(dict[key])
    list.sort()
    for i in range(10):
	wc = list[i]
        print "%s\t%d" % (wc.word, wc.count)

if (len(sys.argv) != 2):
   usage()
wordUse(sys.argv[1])

File: src/oneShot/testScripters/wordUse/wordUse.c
/* wordUse - Make dictionary of all words and count usages. */
#include "common.h"
#include "linefile.h"
#include "hash.h"


void usage()
/* Explain usage and exit. */
{
errAbort(
  "wordUse - Make dictionary of all words and count usages\n"
  "usage:\n"
  "   wordUse file\n"
  );
}

struct wordCount
/* Count number of uses of word. */
    {
    struct wordCount *next;	/* Next in list. */
    char *word;			/* Allocated in hash. */
    int count;			/* Number of times used. */
    };

int wordCountCmpCount(const void *va, const void *vb)
/* Compare two wordCounts in decreasing order of count. */
{
const struct wordCount *a = *((struct wordCount **)va);
const struct wordCount *b = *((struct wordCount **)vb);
return b->count - a->count;
}

void wordUse(char *fileName)
/* wordUse - Make dictionary of all words and count usages, report top ten. */
{
struct lineFile *lf = lineFileOpen(fileName, TRUE);
char *line, *word;
struct hash *hash = newHash(18);
struct wordCount *wcList = NULL, *wc;
int i;
while (lineFileNext(lf, &line, NULL))
    {
    while ((word = nextWord(&line)) != NULL)
	{
	wc = hashFindVal(hash, word);
	if (wc == NULL)
	    {
	    AllocVar(wc);
	    hashAddSaveName(hash, word, wc, &wc->word);
	    slAddHead(&wcList, wc);
	    }
	wc->count += 1;
	}
    }
slSort(&wcList, wordCountCmpCount);
for (i=0, wc=wcList; i<10 && wc != NULL; wc = wc->next, ++i)
    printf("%s\t%d\n", wc->word, wc->count);
}

int main(int argc, char *argv[])
/* Process command line. */
{
if (argc != 2)
    usage();
wordUse(argv[1]);
return 0;
}
